### ISSUE TITLE
Simple SSD, face detector, training ok with GradientTape BUT loss doesn't decrease if I train an identical model with Model.fit()

### ISSUE THREAD

**Issue by qspi** [keras-team/keras#15546]
**System information**.
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): yes, custom
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Ubuntu 20.04, and also on current Google Colab
- TensorFlow installed from (source or binary): binary, GPU version
- TensorFlow version (use command below): 2.5.1 (2.6 on Google Colab)
- Python version: 3.7
- CUDA/cuDNN version: 11.4, 8.2
- GPU model and memory: Geforce RTX 2080 Ti
- Exact command to reproduce: check my Google colab link below

**Describe the problem**.

Hi, I found a really interesting bug when I was building a simple SSD model to find faces on images.

I have a simple, ResNet-like model, containing only convolutions and max pooling layers. I use the "Faces in the Wild" dataset from UMass. When I create a model and I train it with using GradientTape it learns quickly, and after just 20 epochs it predicts very well face bounding boxes, and validation loss goes below 0.6. On the other hand, if I create the exactly the same model (exactly the same sequential model, using the same Adam optimizer and my same custom loss function), AND I use Model.fit() for training it, it just learns nothing. The validation loss starts from 1.8 and doesn't decrease at all.

I found the issue first on my machine running Tensorflow 2.5.1, but I can reproduce the exactly same behaviour on Google Colab as well, currently running 2.6.0.

**Describe the current behavior**.

No matter I train my model by using the Model.fit() function or a GradientTape, by using the same data and parameters and number of epochs the model should learn the features similarly and predict the face bounding boxes similarly with a similar loss value.

**Describe the expected behavior**.

The Model.fit() and the simple GradientTape training should work identical, both trained model should learn and predict similarly.

**Standalone code to reproduce the issue**.

Please check out the code here in Google Colab:
https://colab.research.google.com/drive/1LvahE1CBuvvxAi7PT0gqcSrFYPDBEJmY?usp=sharing

**Source code / logs**.

Please check out the ends of the verbose outputs of training loops:

Model.fit() version output:
[...]
Epoch 18/20
80/80 [==============================] - 9s 111ms/step - loss: 1.6102 - val_loss: 1.6888
Epoch 19/20
80/80 [==============================] - 9s 111ms/step - loss: 1.6102 - val_loss: 1.6888
Epoch 20/20
80/80 [==============================] - 9s 112ms/step - loss: 1.6102 - val_loss: 1.6888

GradientTape version output:
[...]
epoch: [18/20], Train: [loss: 0.4297], Test: [loss: 0.5801]
epoch: [19/20], Train: [loss: 0.4001], Test: [loss: 0.5757]
epoch: [20/20], Train: [loss: 0.3962], Test: [loss: 0.5701]

Thanks for reading it, let me know what do you think.

**Comment by rchao** [comment]
@qspi did you get a chance to see what the loss values are when you inserted `tf.print` in the loss function?

**Comment by qlzh727** [comment]
Also, on a side note, the loss function usually has a signature for `loss(y_true, y_pred)`, and that's the expectation from the framework. In your code example, the ordering for those two params are reversed, which might be the cause of issue.

**Comment by sachinprasadhs** [comment]
@qspi, as Scott suggested above, your `y_true` and `y_pred` were reversed in both the loss function of model.fit and gradient function, modified it and now I'm able to see the decrease in loss for both and bounding box for both the prediction at final prediction. Please  find the gist [here](https://colab.sandbox.google.com/gist/sachinprasadhs/cd66c4855cc500acd33b2a53e4dc01f2/face_ssd_fit_vs_gradienttape.ipynb#scrollTo=ROTwcSJwAn6Y) for the same. Thanks!

**Comment by qspi** [comment]
Hi @sachinprasadhs , @qlzh727 , sorry for the delay, and thanks a lot, that was my fault, I swapped true and predicted labels. I didn’t answer because found sg related what I’m still investigating… I will open another issue when I have the exact question about that. Anyway, thanks again for your help, and I hope this will help others as well.
